src/log_batch.rs

```diff
@@ -419,18 +422,22 @@ impl LogItemBatch {
     /// The `signature` is both generated by the given `LogFileContext`.
     /// That is, the final checksum of each `LogBatch` consists of this
     /// `signature` and the original `checksum` of the contents.
-    pub(crate) fn prepare_write(buf: &mut Vec<u8>, file_context: &LogFileContext) -> Result<()> {
+    pub(crate) fn prepare_write(
+        &mut self,
+        buf: &mut Vec<u8>,
+        file_context: &LogFileContext,
+    ) -> Result<()> {
         if !buf.is_empty() {
+            let mut footer_checksum = self.checksum;
+            // Insert the signature into the encoded bytes. Rewrite checksum of
+            // `LogItemBatch` in `LogBatch`.
+            let footer_checksum_offset = buf.len() - LOG_BATCH_CHECKSUM_LEN;
             if let Some(signature) = file_context.get_signature() {
-                // Insert the signature into the encoded bytes. Rewrite checksum of
-                // `LogItemBatch` in `LogBatch`.
-                let footer_checksum_offset = buf.len() - LOG_BATCH_CHECKSUM_LEN;
-                let original_checksum = codec::decode_u32_le(&mut &buf[footer_checksum_offset..])?;
                 // The final checksum is generated by `signature` ***XOR***
                 // `original checksum of buf`.
-                (&mut buf[footer_checksum_offset..])
-                    .write_u32::<LittleEndian>(original_checksum ^ signature)?;
+                footer_checksum ^= signature;
```

Comment:
No modification on `self.checksum`.
`self.checksum` just caches the local footer checksum of `LogItemBatch`. The final signature just use the `self.checksum ^ signature`. And on the 2nd time, when `engine.write(...)` calls `prepare_write(...)` again, it will be replaced by `self.checksum ^ signature2`.

Reply:
Okay, you need to change the signature from `&mut self` to `&self`.

Reply:
> Okay, you need to change the function signature from `&mut self` to `&self`.

Yep, I found that I've missed it.